### ISSUE TITLE
How to add ":authority" header to gRPC call?

### ISSUE THREAD

**Issue by sandersaares** [improbable-eng/grpc-web#266]
I need to add the ":authority" HTTP2 header for my gRPC calls to be correctly routed to my gRPC server. This is exposed as the `grpc.default_authority` channel option in gRPC core.

I believe this would be configured in the channel options of the grpcwebproxy component, right? Is there currently support for configuring this? If not, I request it be added.

For background, see https://github.com/grpc/grpc/issues/14900

**Comment by johanbrandhorst** [comment]
I don't think we support this right now, but it certainly seems a worthwhile addition. I think there's a few ways we could manage this:

1. Support the [gRPC Name resolution syntax](https://github.com/grpc/grpc/blob/master/doc/naming.md#name-syntax). Particularly for DNS I think this means something like `dns://<authority>/<address>`. We could support this in the `host` parameter in client initialization and populate the `host` or `authority` HTTP headers in the request sent to the proxy.
2. Allow metadata headers like `host` or perhaps even `:authority` to be specified and translated in the proxy.

Thinking a little more about it, we may need to have this discussion in the larger context of the [gRPC-Web protocol](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-WEB.md) itself. It doesn't specify anything about the `authority` header specifically, though it should probably be reserved.

Do you think it's worth opening an issue against grpc/grpc again?

**Comment by sandersaares** [comment]
I am entirely unfamiliar with the details of the gRPC-Web protocol so I am unable to comment much on that aspect. I do notice the gRPC name resolution syntax appears to use the word "authority" in a completely different sense (=DNS server), so I would avoid mixing and "authority" in a DNS context with the ":authority" HTTP2 header.

For my real world scenarios, having some `--backend_authority=xyz` switch on grpcwebproxy (that ends up calling [DialOption.WithAuthority](https://godoc.org/google.golang.org/grpc#WithAuthority)) seems like an ideal solution. In essence, the authority is just an extra specifier added to the service address specified via `--backend_addr`. I would treat these two pieces of data the same.

**Comment by johanbrandhorst** [comment]
I see, yes, that makes sense too. It sounds like it shouldn't be too hard to implement. What can I do to help you get started with a PR to get this in?

**Comment by sandersaares** [comment]
Good question. My knowledge of gRPC usage from Go is minimal - I have only used the .NET API so far. I do not see any obvious place where DialOptions are configured in grpcwebproxy. Can you point out to me where the right place in the code to even configure them would be?

**Comment by johanbrandhorst** [comment]
First job would be to add a flag to the `grpcwebproxy` program:

https://github.com/improbable-eng/grpc-web/blob/2974dfa8ecb5f854fed195463620781895a033af/go/grpcwebproxy/main.go#L29

Second would be to use that flag when making the call to the backend. You'll need to add a parameter to configure the authority to this call:

https://github.com/improbable-eng/grpc-web/blob/2974dfa8ecb5f854fed195463620781895a033af/go/grpcwebproxy/main.go#L50

Then we'll probably want to refactor this line 

https://github.com/improbable-eng/grpc-web/blob/2974dfa8ecb5f854fed195463620781895a033af/go/grpcwebproxy/main.go#L128

a little so that you can specify the authority dial option only if the authority is actually set. Does that make sense?

**Comment by johanbrandhorst** [comment]
Sorry, that's slightly wrong. The call that you need to pass the authority configuration to is

https://github.com/improbable-eng/grpc-web/blob/2974dfa8ecb5f854fed195463620781895a033af/go/grpcwebproxy/backend.go#L46

**Comment by sandersaares** [comment]
I wonder if my first thought at the solution might be misguided. The obvious drawback with adding a parameter for this is that to make calls with multiple `:authority` values, one would need multiple proxies. While perhaps not a deal-killer, it does rather feel like needless complexity.

In this light your original reply's point 2 seems to be an attractive approach - to have the authority string come as input for the proxy, perhaps as a metadata header as you suggest. It would need to be a separate value from the host, as the pattern would be that the proxy references the gateway as the host, which then uses the authority string to route it to the right server instance.

I am currently not yet familiar with gRPC-Web and plan to explore it over the coming weeks, so I am not currently in a position to form an opinion of how exactly this might happen. I see the gRPC-Web protocol link you posted is in the main grpc repository - would that be the place to have this discussion? Would you be willing to initiate it there? I fear I do not have a sufficient understanding of the gRPC-Web stack to consider any of the gRPC-Web relevant angles yet.

**Comment by sandersaares** [comment]
In my scenario, there is one backend server (in the sense of gRPC connections going to one IP address) that implements different services among which the client selects the proper set using the `:authority` header. If a regular (non-web) gRPC client connects to my services, it always uses the same host/IP but a different authority value to access different sets of functionality inside.

If my limited understanding of gRPC-Web is correct, this would not require multiple instances of grpcwebproxy if the latter would simply forward the `:authority` header from the client.

Does this description make sense?

**Comment by johanbrandhorst** [comment]
I'm not terribly familiar with the definition of the `authority` header, is this kind of in-service proxying part of the gRPC http2 spec? I'd imagine the clients would otherwise select service based on routing, i.e. `/myproto.Service1/Service1Method`, `/myproto.Service2/Service2Method`. Is `authority` necessary here?

**Comment by sandersaares** [comment]
Path-based routing was also my first assumption but when I tried it, it did not work for some reason that was not obvious to me (the requests never arrived on the server IIRC).

I originally posted an issue in grpc/grpc asking for path routing support (https://github.com/grpc/grpc/issues/14900) but was told to use the authority header instead. Why this is preferred I do not know - I just know that it works quite well, so I would happily wish to extend this pattern to my usage of gRPC web clients, as well.

**Comment by johanbrandhorst** [comment]
I'm using the paths generated from the proto definitions relative to the root of the hosting server - any preceding paths will indeed not work. Anyway, it does seem like the `authority` header is a well-known header in the spec, so we should probably handle it.

For now, have you tried manually setting the authority header in the request metadata? It might already work.

**Comment by sandersaares** [comment]
I have not yet tried setting it manually, no - right now I am looking at the documentation and will work on the experimental implementation in November most likely.

I see what you mean by paths. The level of routing I do with the authority header is a level above that. So it would be something like the equivalent of `/Foo/myproto.Service1/Service1Method` and `/Foo/myproto.Service2/Service2Method` and `/Bar/myproto.Service3/Service3Method`.

How it is used is that my gRPC backend server is just a gateway that selects which real gRPC server to use based on the authority (as it has no knowledge of which actual services are implemented by which server).

**Comment by johanbrandhorst** [comment]
So you've got a manual grpc backend proxy that looks at the authority header and routes to the correct gRPC client (connected to another gRPC server) based on that?

**Comment by sandersaares** [comment]
No, the gRPC proxying based on authority is a built-in feature in the nginx web server.

**Comment by johanbrandhorst** [comment]
Ah, right, I see. Yes that makes sense then. I imagine you might be able to just put the header in the metadata in the frontend and it would "just work".

**Comment by sandersaares** [comment]
I tried putting it in the metadata and confirm that it did not just work - the proxy seems to treat it as just a gRPC metadata key-value pair and does not transform it into the HTTP2 header.
